File: tests/diag.py
# Copyright 2021 NVIDIA Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import numpy as np

import legate.numpy as lg


def test():
    for k in [0, -1, 1, -2, 2]:
        print(f"diag(k={k})")
        a = lg.array(
            [
                [1, 2, 3, 4],
                [5, 6, 7, 8],
                [9, 10, 11, 12],
                [13, 14, 15, 16],
                [17, 18, 19, 20],
            ]
        )
        an = np.array(
            [
                [1, 2, 3, 4],
                [5, 6, 7, 8],
                [9, 10, 11, 12],
                [13, 14, 15, 16],
                [17, 18, 19, 20],
            ]
        )

        b = lg.diag(a, k=k)
        bn = np.diag(an, k=k)
        assert np.array_equal(b, bn)

        c = lg.diag(b, k=k)
        cn = np.diag(bn, k=k)
        assert np.array_equal(c, cn)

        d = lg.array(
            [
                [1, 2, 3, 4, 5],
                [6, 7, 8, 9, 10],
                [11, 12, 13, 14, 15],
                [16, 17, 18, 19, 20],
            ]
        )
        dn = np.array(
            [
                [1, 2, 3, 4, 5],
                [6, 7, 8, 9, 10],
                [11, 12, 13, 14, 15],
                [16, 17, 18, 19, 20],
            ]
        )

        e = lg.diag(d, k=k)
        en = np.diag(dn, k=k)
        assert np.array_equal(e, en)

        f = lg.diag(e, k=k)
        fn = np.diag(en, k=k)
        assert np.array_equal(f, fn)

    return


if __name__ == "__main__":
    test()

File: src/matrix/diag.cc
/* Copyright 2021 NVIDIA Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "matrix/diag.h"
#include "matrix/diag_template.inl"

namespace legate {
namespace numpy {

using namespace Legion;

template <LegateTypeCode CODE>
struct DiagImplBody<VariantKind::CPU, CODE> {
  using VAL = legate_type_of<CODE>;

  void operator()(const AccessorWO<VAL, 2> &out,
                  const AccessorRO<VAL, 1> &in,
                  coord_t distance,
                  const Point<2> &start_out,
                  coord_t start_in) const
  {
    for (coord_t idx = 0; idx < distance; ++idx)
      out[start_out[0] + idx][start_out[1] + idx] = in[start_in + idx];
  }

  void operator()(const AccessorWO<VAL, 1> &out,
                  const AccessorRO<VAL, 2> &in,
                  coord_t distance,
                  coord_t start_out,
                  const Point<2> &start_in) const
  {
    for (coord_t idx = 0; idx < distance; ++idx)
      out[start_out + idx] = in[start_in[0] + idx][start_in[1] + idx];
  }

  void operator()(const AccessorRD<SumReduction<VAL>, true, 1> &out,
                  const AccessorRO<VAL, 2> &in,
                  coord_t distance,
                  coord_t start_out,
                  const Point<2> &start_in) const
  {
    for (coord_t idx = 0; idx < distance; ++idx)
      out.reduce(start_out + idx, in[start_in[0] + idx][start_in[1] + idx]);
  }
};

void deserialize(Deserializer &ctx, DiagArgs &args)
{
  deserialize(ctx, args.extract);
  deserialize(ctx, args.needs_reduction);
  deserialize(ctx, args.k);
  deserialize(ctx, args.shape);
  deserialize(ctx, args.out);
  deserialize(ctx, args.in);
}

/*static*/ void DiagTask::cpu_variant(const Task *task,
                                      const std::vector<PhysicalRegion> &regions,
                                      Context context,
                                      Runtime *runtime)
{
  diag_template<VariantKind::CPU>(task, regions, context, runtime);
}

namespace  // unnamed
{
static void __attribute__((constructor)) register_tasks(void) { DiagTask::register_variants(); }
}  // namespace

}  // namespace numpy
}  // namespace legate
